### DISCUSSION TITLE
How to Hide Other Elements In Same Foreach Loop?

### DISCUSSION THREAD

**Discussion by reasecret** [alpinejs/alpine#1330] [1. Help] [answered]
I've a component like this. If I click edit icon an input appears next to label (Like test 2)

![resim](https://user-images.githubusercontent.com/32717963/119032257-a77a6400-b9b4-11eb-805d-1e8c699872d6.png)

I want to do this: If I click any of other pen icon, other inputs should be hidden. Inputs are dynamic and count can be change. I'm new at Alpine Js and can't find examples like this. How can I do this with this component: 

```
@foreach(Auth::user()->educations as $key => $education)
            <div x-data="{ showInput: false, educationId: '' }" class="flex items-center p-4">
                <div class="flex justify-between items-center w-full">
                    <div class="flex items-center justify-start">
                       <span class="bg-gray-100 py-2 px-4 rounded-l">
                          {{ $education->name }}
                       </span>
                        <input x-show="educationId === '{{ $education->id }}'" class="form-input p-1 ml-4" type="text" wire:model="educationOldName" value="{{ $education->name }}">
                    </div>
                </div>
                <div class="flex items-center">
                    <svg x-show="educationId === '{{ $education->id }}'" wire:click="updateEducation({{ $education->id }})" data-tippy-content="Save" class="cursor-pointer"/></svg>
                    <svg data-tippy-content="Edit" @click="educationId = '{{ $education->id }}'" class="cursor-pointer" /></svg>
                    <svg data-tippy-content="Delete" wire:click="removeEducationFromUser({{ $education->id }})" class="cursor-pointer" /></svg>
                </div>
            </div>
        @endforeach
```

**Comment by gwleuverink** [comment]
You could wrap all inputs in a alpine component and move the logic of keeping track of which item is currently being edited in there. 

But since I see you're using livewire, so you might also opt to do it via the back end. It might make the experience a bit less snappy, but can simplify it a lot.

**Reply by reasecret** [reply]
Could you give me some code for example? I can't make logic for this.

**Reply by gwleuverink** [reply]
Will something like this do the trick? 

``` html
<div
    x-data="{
        activeIndex: null,

        edit(index) {
            this.activeIndex = index

            this.$nextTick(() => {
                this.$refs['input-' + index].focus()
            });
        }
    }"
    class="max-w-sm flex flex-col space-y-3"
>

    @foreach ($educations as $education)

        <div class="flex items-center space-x-4">

            <span>title</span>

            <span class="flex-grow">
                <input
                    x-show="activeIndex === {{ $loop->index }}"
                    x-ref="input-{{ $loop->index }}"
                    type="text"
                    class="w-full border border-gray-300 px-2 py-1" />
            </span>

            <button x-on:click="edit({{ $loop->index }})" type="button">edit</button>

        </div>

    @endforeach

</div>
```

**Comment by gwleuverink** [comment, marked as answer]
Since You're using Livewire you could also opt to go a different way by modelling that behavior with php. It comes down to personal preference so really there is not one 'correct' way of doing things. Howerver, I like to offload these things to Livewire if I can get away with it. This way you don't need any javascript at all and there is only one single source of truth in terms of state.

Here's a simplified example:

``` php
class EducationManager extends Component
{
    public ?Education $editing = null;

    public function edit(Education $education)
    {
        // TODO: Authorize if education can be edited by this user

        $this->editing = $education;
    }

    public function save()
    {
        // TODO: Authorize if education can be saved by this user

        $this->validate();

        $this->editing->save();

        $this->reset('editing');
    }

    protected $rules = [
        'editing.title' => ['required', 'min:4']
    ];

    public function render()
    {
        return view('livewire.education-manager', [
            'educations' => auth()->user()->educations
        ]);
    }
}

```

``` html
<!-- education-manager.blade.php -->

<div class="max-w-sm flex flex-col space-y-3">
    
    @foreach ($educations as $education)

        <div class="flex items-center space-x-4">

            <span>{{ $education->title }}</span>

            <div class="flex-grow">

                @if($editing && $education->id === $editing->id)
                    <form wire:submit.prevent="save" novalidate autocomplete="off" class="flex">

                        <input wire:model="editing.title" type="text" class="w-full border border-gray-300 px-2 py-1" />
                        <button type="submit">save</button>

                    </form>

                    <span class="text-red-500">
                        @error('editing.title') {{ $message }} @enderror
                    </span>
                @endif

            </div>

            <button wire:click.prefetch="edit({{ $education->id }})">edit</button>

        </div>

    @endforeach

</div>
```

Look at that! Completely dynamic without using Javascript. There is some slowdown in terms of UI  responsiveness, since toggling an active input requires a roundtrip to the server. Though this slowdown is greatly reduced by using `wire:click.prefetch`.

**Reply by reasecret** [reply]
Thank you for your replies. Second method worked fine but there is a one little issue. If I click edit icon and then click another icon, can't click icon second time. Icon's tooltip is not showing also. Here is a gif:

![save](https://user-images.githubusercontent.com/32717963/119235507-8b54ff00-bb3b-11eb-8361-367c016e741a.gif)

But if I remove .prefetch there is no problem. I'm OK with that and for now this is not a big issue for me. I'll use like this. Thank you a lot for your effort. :)

**Reply by gwleuverink** [reply]
No problem, always happy to help.

It is a weird bug. I suspect it could be a dom diffing issue. Maybe you can try to add a unique wire:key to the topmost element in the loop?

Check out this section of the docs. https://laravel-livewire.com/docs/2.x/troubleshooting
